### PR TITLE
feat(core): add new text-shadow property

### PR DESCRIPTION
**Pull request by tralves**
## PR Checklist

- [X] The PR title follows our guidelines: https://github.com/NativeScript/NativeScript/blob/master/CONTRIBUTING.md#commit-messages.
- [X] There is an issue for the bug/feature this PR is for. To avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it.
- [X] You have signed the [CLA](http://www.nativescript.org/cla).
- [X] All existing tests are passing: https://github.com/NativeScript/NativeScript/blob/master/DevelopmentWorkflow.md#running-unit-tests.
- [ ] Tests for the changes are included - https://github.com/NativeScript/NativeScript/blob/master/WritingUnitTests.md.

## What is the current behavior?
NativeScript does not support `text-shadow`. This has been a requested feature (#3597, #5268).

## What is the new behavior?
Adds `text-shadow` css property (and `textShadow` attribute).

It implements a subset of the [text-shadow web syntax](https://developer.mozilla.org/en-US/docs/Web/CSS/text-shadow). The expected format is:
```
<offset-x> <offset-y> <blur-radius> <color>
e.g.
2 10 4 rgb(255, 100, 100)
2px 10px 2px rgba(10, 10, 10, 0.5)
1 1 1 #55a
2 2 2 #aaa
``` 
The reasons for this limitation are:
- some of the default values in the web syntax are not clear (i.e., people shouldn't use them anyway);
- It is impossible to implement multiple shadows (e.g. `1px 1px 2px red, 0 0 1em blue, 0 0 0.2em blue;`);
- parsing/tokenisation would not be trivial because there are multiple formats for colors (`rba(0, 0, 0)`, `#fff`, etc.). It would be a lot of work;

I thought I should have more feedback on this PR before adding more work.

I think it PR would still need:
- Unit tests (I didn't add them yet, but I added a demo for this feature in the `apps.ui`);
- Add more syntax options (maybe? which ones?);
- Validation/syntax errors;
- Support for formatted strings and spans (not sure if it's even possible);
- Support for tab headers (not sure if it's even possible);

Demo (apps.ui):

ns-text-shadows

Fixes/Implements/Closes #[Issue Number].
#3597 and part of #5268.

### CODE REVIEW

**Comment by tralves** [comment]
Hi!
I see that this is on the roadmap :D. As I said this PR needs some polish. Do you have pointers or tips to what I should do to make your lives easier?

File: packages/core/ui/text-base/text-base-common.ts [hunk 1]
```diff
@@ -26,26 +27,6 @@ export abstract class TextBaseCommon extends View implements TextBaseDefinition
...
-	 * NS Core uses nativeViewProtected for everything related to layout, so that it can be measured, added to the parent view as a child, ect.
-	 *
-	 * However, its internal view would be the actual TextView/TextInput and to allow that sub-view to have the normal TextView/TextInput
-	 * class features, which we expose and to allow them to work on it, the internal TextView/TextInput is what the needs to have the class values applied to it.
-	 *
-	 * So all code that works on what is expected to be a TextView/TextInput should use `nativeTextViewProtected` so that any third party
-	 * components that need to have two separate components can work properly without them having to duplicate all the TextBase (and decendants) functionality
-	 * by just overriding the nativeTextViewProtected getter.
-	 **/
 	get nativeTextViewProtected() {
```
Comment (farfromrefug) [inline packages/core/ui/text-base/text-base-common.ts:30]: Why did you remove that comment ? It is actually an important comment

File: packages/core/ui/text-base/index.android.ts [hunk 2]
```diff
@@ -378,6 +378,19 @@ export class TextBase extends TextBaseCommon {
...
 	}
 
+	[textShadowProperty.getDefault](value: number) {
+		return {
+			radius: this.nativeTextViewProtected.getShadowRadius(),
+			offsetX: this.nativeTextViewProtected.getShadowDx(),
+			offsetY: this.nativeTextViewProtected.getShadowDy(),
+			color: this.nativeTextViewProtected.getShadowColor(),
+		};
+	}
```
Comment (farfromrefug) [inline packages/core/ui/text-base/index.android.ts:388]: First I dont think we need a default here. We have to be careful. Get default are slow because they make native calls. And yours even make 4 of them.
We now a default view as no shadow so not having a default value should be fine and faster.
Also you should use a const value to access native view protected instead of accessing it 4 times. If a class define it as a getter your would unnecessarly call it 4 times

File: packages/core/ui/text-base/index.ios.ts [hunk 3]
```diff
@@ -188,6 +188,10 @@ export class TextBase extends TextBaseCommon {
 		this._setNativeText();
 	}
 
+	[textShadowProperty.setNative](value: TextShadow) {
+		this._setShadow(value);
```
Comment (farfromrefug) [inline packages/core/ui/text-base/index.ios.ts:192]: Why create a _setShadow method If you call it only once?
It makes the class."bigger" for now reason

Comment (NathanaelA) [reply packages/core/ui/text-base/index.ios.ts:192]: Doesn't really make the class bigger (two lines?), that code has to live somewhere.     I personally prefer it to be a separate function as it makes it easier to work on shadow stuff if we detect stuff broken.  In addition we might decide later to push it up the stack in a base class for other reasons.   Finally v8 engine typically optimizes smaller functions better than larger functions (& there used to be a limit to function sizes v8 will optimize, not sure if it is still present).  So having this in a separate function makes a lot of sense to me...

File: packages/core/ui/text-base/text-base-common.ts (hunk 1 above)
Comment (NathanaelA) [reply packages/core/ui/text-base/text-base-common.ts:30]: Agreed, this comment should remain, it is important to understand why their is two separate getters that normally point to the same property.

**Comment by farfromrefug** [comment]
@NathanWalker you merged this without review comments being fixed ? that PR removes things which should not have been touched here !

**Comment by rigor789** [comment]
@farfromrefug it's on the `release/8.0.0` branch where we are still going to do cleanups & changes. Just helps manage things together - and mainly helps resolving any conflicts early.

**Comment by farfromrefug** [comment]
I get it. just that PR removes comments which should remain there for everyone to understand things. 
I don't see how you are going to bring them back now (except manually obviously) now that it is merged.
thanks for your work guys !
